File: src/sort/merge-sort.ts
import { swap, left, right } from "./util"
import { Sortable } from "../model/Sortable";


export class MergeSort implements Sortable {

    merge = (arr: number[], l: number, m: number, r: number) => {
        let leftArr = [];
        let rightArr = [];

        for (let index = l; index <= m; index++) {
            leftArr.push(arr[index]);
        }
        leftArr.push(Number.MAX_SAFE_INTEGER);

        for (let index = m + 1; index <= r; index++) {
            rightArr.push(arr[index]);
        }
        rightArr.push(Number.MAX_SAFE_INTEGER);

        let leftIndex = 0;
        let rightIndex = 0;

        for (let index = l; index <= r; index++) {
            if (leftArr[leftIndex] < rightArr[rightIndex]) {
                arr[index] = leftArr[leftIndex++];
            } else {
                arr[index] = rightArr[rightIndex++];
            }
        }
    }

    mergeSort = (arr: number[], i: number, j: number) => {
        if (i < j) {
            let middle = Math.floor((i + j) / 2);
            this.mergeSort(arr, i, middle);
            this.mergeSort(arr, middle + 1, j);
            this.merge(arr, i, middle, j);
        }
    }

    sort(arr: number[], maxNum?: number) {
        this.mergeSort(arr, 0, arr.length - 1);
    }

}

File: src/main/java/com/github/danimaniarqsoft/sorting/MergeSort.java
package com.github.danimaniarqsoft.sorting;

import java.util.ArrayList;
import java.util.List;

public class MergeSort implements Sort {

    @Override
    public void sort(int[] data) {
        mergeSort(data, 0, data.length - 1);
    }

    private void mergeSort(int[] array, int min, int max) {
        if (min < max) {
            int middle = (int) Math.floor(((double) min + max) / 2);
            mergeSort(array, min, middle);
            mergeSort(array, middle + 1, max);
            merge(array, min, middle, max);
        }
    }

    private static void merge(int[] array, int min, int middle, int max) {

        List<Integer> left = new ArrayList<>();
        List<Integer> right = new ArrayList<>();

        for (int i = min; i <= middle; i++) {
            left.add(array[i]);
        }
        left.add(Integer.MAX_VALUE);

        for (int i = middle + 1; i <= max; i++) {
            right.add(array[i]);
        }
        right.add(Integer.MAX_VALUE);

        int leftIndex = 0;
        int rightIndex = 0;

        for (int i = min; i <= max; i++) {
            if (left.get(leftIndex) < right.get(rightIndex)) {
                array[i] = left.get(leftIndex);
                leftIndex = leftIndex + 1;
            } else {
                array[i] = right.get(rightIndex);
                rightIndex = rightIndex + 1;
            }
        }
    }

}
